### DISCUSSION TITLE
enhance and despeckle

### DISCUSSION THREAD

**Discussion by exekutive** [ImageMagick/ImageMagick#7437] [Help] [answered]
### ImageMagick version

7.1.1

### Operating system, version and so on

any

### Description

I'm trying to find more details about what the -enhance and -despeckle operators do and how they work. The IM documentation wasn't very verbose.

obviously despeckle gets rid of speckles, but how? and what radius of speckles. 1px? 2px? 10px? any color? what contrast threshold does it use if any?

enhance is even more vague. Docs just say "Apply a digital filter to enhance a noisy image."

other operators (-distort for example) are described in extreme academic detail, so this was just a bit disappointing.

**Comment by fmw42** [comment]
I do not know of any real documentation for those noise filters. You would have to read the code. I played around with them and have not found despeckle very useful. Perhaps I have not tried the right kind of noise. It just seems to blur the noise for me.  I found -enhance more useful if you apply it about 5 times. See my examples from my script noisecleaner at http://www.fmwconcepts.com/imagemagick/noisecleaner/index.php

**Reply by exekutive** [reply]
thanks. yeah I noticed similar results. Enhance seems to perform an additional step which increases sharpness or contrast a bit, which despeckle doesn't.

**Comment by snibgo** [comment, marked as answer]
I have sometimes found `-despeckle` effective. The code, in effect.c, has this comment:
```
%  DespeckleImage() reduces the speckle noise in an image while preserving the
%  edges of the original image.  A speckle removing filter uses a complementary
%  hulling technique (raising pixels that are darker than their surrounding
%  neighbors, then complementarily lowering pixels that are brighter than their
%  surrounding neighbors) to reduce the speckle index of that image (reference
%  Crimmins speckle removal).
```
An internet search for "Crimmins speckle removal" finds many pages with explanations, for example https://homepages.inf.ed.ac.uk/rbf/HIPR2/crimmins.htm.

`-enhance` in enhance.c has the vague comment:
```
%  EnhanceImage() applies a digital filter that improves the quality of a
%  noisy image.
```
For ordinary (non-noisy) images, it seems to increase blur in blurry areas, and sharpen sharp edges. A single `-enhance` doesn't do much, but doing three of these has an obvious effect.
